sys/net/application_layer/coapfileserver/coapfileserver.c

```diff
@@ -0,0 +1,420 @@
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+/** Maximum length of an expressible path, including the trailing 0 character. */
+#define COAPFILESERVER_PATH_MAX (64)
+
+/** Constant ETag length */
+#define ETAG_LENGTH 8
+
+/** Data extracted from a request on a file */
+struct requestdata {
+    /** 0-terminated expanded file name in the VFS */
+    char namebuf[COAPFILESERVER_PATH_MAX];
+    uint32_t blocknum2;
+    unsigned int szx2; /* would prefer uint8_t but that's what coap_get_blockopt gives */
+    bool etag_sent;
+    uint8_t etag[ETAG_LENGTH];
+};
+
+/* See stat_etag */
+union stattag {
+    struct stat stat;
+    struct {
+        uint8_t etag[ETAG_LENGTH];
+        /* This will be as many repetitions of ETAG_LENGTH that together
+         * with the original etag field id's larger than the stat */
+        uint8_t etag_padding[(sizeof(struct stat) - 1) / ETAG_LENGTH * ETAG_LENGTH];
+    };
+};
+
+/** Build an ETag based on the given file's VFS stat. If the stat fails,
+ * returns the error and leaves stattag in any state; otherwise there's an etag
+ * in the stattag's field */
+static int stat_etag(char *filename, union stattag *stattag);
+
+/* These are almost but but not quite coap_handler_t -- they require the
+ * request to be already fully read and digested into the last argument. */
+
+static ssize_t coapfileserver_file_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, struct requestdata *request);
+static ssize_t coapfileserver_directory_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, struct requestdata *request);
+/** Create a CoAP response for a given errno (eg. EACCESS -> 4.03 Forbidden
+ * etc., defaulting to 5.03 Internal Server Error) */
+static size_t coapfileserver_errno_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, int err);
+/** Return true if path/name is a directory.
+ *
+ * Currently assembling the values in a buffer, this could be done without a
+ * buffer if an openat / vfs_openat was present.
+ * */
+static bool entry_is_dir(char *path, char *name);
+
+static unsigned _count_char(const char *s, char c)
+{
+    unsigned count = 0;
+    while (*s) {
+        if (*s == c) {
+            ++count;
+        }
+        ++s;
+    }
+
+    return count;
+}
+
+ssize_t coapfileserver_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, void *ctx) {
+    struct coapfileserver_entry *entry = ctx;
+    struct requestdata request = {
+        .etag_sent = false,
+        .blocknum2 = 0,
+        .szx2 = CONFIG_NANOCOAP_BLOCK_SIZE_EXP_MAX,
+    };
+    /** Index in request.namebuf. Must not point at the last entry as that will be
+     * zeroed to get a 0-terminated string. */
+    size_t namelength = 0;
+    bool trailing_slash = false;
+    /** If no path component comes along at all, it'll count as a trailing
+     * slash no matter the trailing_slash value */
+    bool any_component = false;
+    uint8_t errorcode = COAP_CODE_INTERNAL_SERVER_ERROR;
+    uint8_t strip_remaining = _count_char(entry->resource, '/');
```

Comment:
if `entry->resource` has a `/` at the end, then one `URI_PATH` option is skipped too many.

Reply:
I want to get rid of this entirely with #17957, until then just make sure the resource matches the CoAP resource (and those should not have a trailing slash, right @chrysn?)

Reply:
CoAP resources can easily have trailing slashes (expressed as empty Uri-Path options) -- in the ->resource entry the requirement was not to have one (to "pass on the slash onto the file system", figuratively speaking). For [COAP_MATCH_SUBTREE](https://api.riot-os.org/group__net__nanocoap.html#ga8c87dddfffd38c7c88c32b49fad9cb26) either might be acceptable, but this handler would express the requirement not to have one.